### PR TITLE
Fix #3318: Add check to ensure all todos are addressed corresponding to the closed issue

### LINKED ISSUE #3318: Create a workflow that verifies (through a Kotlin Script) that all the TODOs are addressed for a closed issue
We want to make sure that no issue is closed without its TODOs being addressed. If it is closed without the addressing of all its TODOs, then we want it to reopen again along with a comment giving the context of why it was reopened.

### PR DESCRIPTION
**Pull request by Sparsh1212**
## Explanation
Fixes #3318

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Comment by Sparsh1212** [comment]
@BenHenning PTAL at this issue as a POC for this script.
Ignore the other issues on this repo as they were for debugging purposes.
https://github.com/Sparsh1212/oppia-android-clone/issues/2

TLDR: Actual behavior was the same as the expected behavior for all cases.

The flow of testing the script was as follows:
- At first, there were no todos in the codebase for this issue, so when I closed this, this wasn't reopened and no comment was posted.
- Then I reopened it manually to check if the workflow passes and nothing else happens, the actual behavior was the same as expected.
- Then I added some todos in the repo corresponding to this issue via a commit and then closed this issue. The actual result was: The issue was reopened automatically by the GitHub actions bot and a comment was also posted mentioning the TODO list.
- I again closed the issue to check if a duplicate comment is being posted or not, and the result was that now it was gained reopened by the action bot but this time no comment was added.
- The final step was to add more commits to the codebase which adds more todos for this issue. I again closed the issue, the result was it got reopened and a new comment was added since the todo list got changed.

Edit: Another thing I did, as per our discussion in the meet was to remove all the TODOs for the TODO Issue, and the outcome was, the workflow was successful on closing it and it wasn't reopened again automatically.

**Comment by Sparsh1212** [comment]
@BenHenning 
I have not added the `ClosedIssueCheckTest` todo exemptions to the exemptions list yet, as the PR is in review for the first time. Will update it after the first pass. So the TODO failing CI is actually expected.

**Comment by BenHenning** [comment]
Hi. Will need to look at this tomorrow.

File: .github/workflows/closed_issue_check.yml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Contains job to check that all the todos are resolved for the closed issue.
+
+name: Closed Issue Check
```
Comment (BenHenning) [inline .github/workflows/closed_issue_check.yml]: I suggest instead framing the name, documentation, and file to be based around issue workflows, instead. One specific example is for closed issues, but we may add more in the future (generic buckets for actions workflows are preferred over more specific ones).

File: .github/workflows/closed_issue_check.yml [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+on:
+  issues:
+    types: [closed, reopened]
+
+permissions:
+  issues: write
+
+jobs:
+  script_check:
+    name: Script Check
```
Comment (BenHenning) [inline .github/workflows/closed_issue_check.yml]: Should be renamed to be specific to the check in question.

File: .github/workflows/closed_issue_check.yml [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+          gh issue reopen ${{ github.event.issue.number }}
+
+      - name: Check for duplicate comment
+        id: duplicateCommentCheck
+        if: failure() && steps.closedIssueCheck.outcome == 'failure'
+        env:
+          GITHUB_TOKEN: ${{ github.token }}
+        run: |
+          gh issue view ${{ github.event.issue.number }} --json comments --jq '.comments[-1].body' > $(pwd)/latest_comment.txt
+          bazel run //scripts:comment_check -- $(pwd) latest_comment.txt todo_list.txt
```
Comment (BenHenning) [inline .github/workflows/closed_issue_check.yml]: Is todo_list.txt the correct name for this file? It seems more like it contains the contents for an issue comment.

File: scripts/BUILD.bazel [hunk 4]
```diff
@@ -163,3 +163,21 @@ kt_jvm_binary(
         "//scripts/src/java/org/oppia/android/scripts/todo:todo_check_lib",
     ],
 )
+
+kt_jvm_binary(
+    name = "closed_issue_check",
```
Comment (BenHenning) [inline scripts/BUILD.bazel]: This should be named specifically for TODOs (ditto for the class & Bazel target names).

File: scripts/BUILD.bazel [hunk 5]
```diff
@@ -163,3 +163,21 @@ kt_jvm_binary(
...
+    testonly = True,
+    main_class = "org.oppia.android.scripts.todo.ClosedIssueCheckKt",
+    runtime_deps = [
+        "//scripts/src/java/org/oppia/android/scripts/todo:closed_issue_check_lib",
+    ],
+)
+
+kt_jvm_binary(
+    name = "comment_check",
```
Comment (BenHenning) [inline scripts/BUILD.bazel]: Ditto here--this is specific to TODO issues.

File: scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+ *   bazel run //scripts:closed_issue_check -- <path_to_directory_root> <closed_issue_number>
+ *   <github_sha>
+ *
+ * Arguments:
+ * - path_to_directory_root: directory path to the root of the Oppia Android repository.
+ * - closed_issue_number: issue number of the closed issue.
+ * - github_sha: sha of the latest commit on the develop branch.
+ *
+ * Example:
+ *   bazel run //scripts:closed_issue_check -- $(pwd) 6 3898188
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt]: Small nit, but maybe use a real hash of the develop branch (can be shortened).

File: scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  closedIssueFailureTodos: List<Todo>,
+  githubPermalinkUrl: String
+) {
+  val todoListFile = File(repoPath + "todo_list.txt")
+  todoListFile.appendText("The issue is reopened because of the following unresolved TODOs:")
+  todoListFile.appendText("\n")
+  closedIssueFailureTodos.sortedWith(compareBy({ it.filePath }, { it.lineNumber }))
+    .forEach { todo ->
+      todoListFile.appendText(
+        "$githubPermalinkUrl${(todo.filePath).removePrefix(repoPath)}#L${todo.lineNumber}"
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt]: Nit, but I suggest adding the ``/`` here instead of the URL variable so that this string reads a bit more like a URL.

File: scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+      )
+      todoListFile.appendText("\n")
+    }
+}
+
+/**
+ * Logs the Closed Issue Check failures.
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt]:
```suggestion
 * Logs the closed issue check failures.
```

File: scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt [hunk 9]
```diff
@@ -0,0 +1,227 @@
...
+      File("${retrieveTestFilesDirectoryPath()}/todo_list.txt").useLines { it.toList() }
+
+    assertThat(exception).hasMessageThat().contains(CLOSED_ISSUE_CHECK_FAILED_OUTPUT_INDICATOR)
+    val failureMessage =
+      """
+      The following TODOs are unresolved for the closed issue:
+      - ${retrieveTestFilesDirectoryPath()}/TempFile1.kt:1
+      - ${retrieveTestFilesDirectoryPath()}/TempFile2.bazel:3
+      - ${retrieveTestFilesDirectoryPath()}/TempFile3.xml:1
+      - ${retrieveTestFilesDirectoryPath()}/TempFile3.xml:4
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt]: This already seems verified in the previous test--suggest omitting it here & focusing on the outcome specific to the behavior corresponding to this test.

File: scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt [hunk 10]
```diff
@@ -0,0 +1,227 @@
...
+      """.trimIndent()
+    tempFile1.writeText(testContent1)
+    tempFile2.writeText(testContent2)
+    tempFile3.writeText(testContent3)
+
+    val exception = assertThrows(Exception::class) {
+      main(retrieveTestFilesDirectoryPath(), "169877", "abmzuyt")
+    }
+    val fileContentList =
+      File("${retrieveTestFilesDirectoryPath()}/todo_list.txt").useLines { it.toList() }
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt]:
```suggestion
      File("${retrieveTestFilesDirectoryPath()}/todo_list.txt").readLines()
```

File: scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+    )
+    assertThat(fileContentList.elementAt(2)).isEqualTo(
+      "https://github.com/oppia/oppia-android/blob/abmzuyt/TempFile2.bazel#L3"
+    )
+    assertThat(fileContentList.elementAt(3)).isEqualTo(
+      "https://github.com/oppia/oppia-android/blob/abmzuyt/TempFile3.xml#L1"
+    )
+    assertThat(fileContentList.elementAt(4)).isEqualTo(
+      "https://github.com/oppia/oppia-android/blob/abmzuyt/TempFile3.xml#L4"
+    )
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt]:
```suggestion
    assertThat(fileContentList).containsExactly(
      "The issue is reopened because of the following unresolved TODOs:",
      "https://github.com/oppia/oppia-android/blob/abmzuyt/TempFile1.kt#L1",
      "https://github.com/oppia/oppia-android/blob/abmzuyt/TempFile2.bazel#L3",
      "https://github.com/oppia/oppia-android/blob/abmzuyt/TempFile3.xml#L1",
      "https://github.com/oppia/oppia-android/blob/abmzuyt/TempFile3.xml#L4"
    ).inOrder()
```

Bit cleaner & idiomatic.

File: scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+
+  for (index in 1 until latestCommentContentList.size) {
+    // The commit SHA can vary from workflow to workflow. This can make the permalinks different.
+    // Hence, we are comparing by the relative file path.
+    val latestCommentLineContent = latestCommentContentList[index].substring(85)
+    val failureCommentLineContent = failureCommentContentList[index].substring(85)
+    if (latestCommentLineContent != failureCommentLineContent) return
+  }
+
+  throw Exception("LATEST COMMENT IS SAME AS THE FAILURE COMMENT")
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt]: This seems backwards--could we use a passing indicator for success instead of a failing indicator?

File: scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+  val failureCommentContentList = File(repoPath, failureCommentFilePath).readText().trim().lines()
+
+  if (latestCommentContentList.size != failureCommentContentList.size) return
+
+  if (latestCommentContentList.first() != failureCommentContentList.first()) return
+
+  for (index in 1 until latestCommentContentList.size) {
+    // The commit SHA can vary from workflow to workflow. This can make the permalinks different.
+    // Hence, we are comparing by the relative file path.
+    val latestCommentLineContent = latestCommentContentList[index].substring(85)
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt]: The '85' is a magic value. Can we use something more deterministic to ignore the hash?

File: scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+
+import com.google.common.truth.Truth.assertThat
+import org.junit.Rule
+import org.junit.Test
+import org.junit.rules.TemporaryFolder
+import org.oppia.android.testing.assertThrows
+
+/** Tests for [CommentCheck]. */
+class CommentCheckTest {
+  private val permalinkPrefix = "https://github.com/oppia/oppia-android/blob/"
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt]: Ditto here--suggest moving the last ``/`` to the later comments for slightly better readability (super nit).

File: scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+  @Test
+  fun testCommentCheck_emptyLatestCommentBody_commentCheckShouldPass() {
+    val latestCommentFile = tempFolder.newFile("latest_comment.txt")
+    val scriptFailureCommentFile = tempFolder.newFile("todo_list.txt")
+    val latestCommentContent = ""
+    val scriptFailureCommentContent =
+      """
+      The issue is reopened because of the following unresolved TODOs:
+      $permalinkPrefix$dummySha1/scripts/BUILD.bazel#L184
+      $permalinkPrefix$dummySha1/scripts/BUILD.bazel#L185
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt]: Could we output these as a markdown list, instead, or does that mess up the automatic previewing that GitHub does?

File: scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+  @Rule
+  @JvmField
+  var tempFolder = TemporaryFolder()
+
+  /**
+   * This is the case when no comment is present on the issue. In this case the Github API results
+   * in an empty comment body.
+   */
+  @Test
+  fun testCommentCheck_emptyLatestCommentBody_commentCheckShouldPass() {
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt]: Nice test case.

File: scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt [hunk 17]
```diff
@@ -0,0 +1,49 @@
...
+
+  val failureCommentContentList = File(repoPath, failureCommentFilePath).readText().trim().lines()
+
+  if (latestCommentContentList.size != failureCommentContentList.size) return
+
+  if (latestCommentContentList.first() != failureCommentContentList.first()) return
+
+  for (index in 1 until latestCommentContentList.size) {
+    // The commit SHA can vary from workflow to workflow. This can make the permalinks different.
+    // Hence, we are comparing by the relative file path.
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt]: Can you please also add a test case for this scenario to make sure it's working as expected?

File: scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+      $permalinkPrefix$dummySha1/scripts/BUILD.bazel#L184
+      $permalinkPrefix$dummySha1/scripts/BUILD.bazel#L185
+      
+      
+      """.trimIndent()
+    val scriptFailureCommentContent =
+      """
+      The issue is reopened because of the following unresolved TODOs:
+      $permalinkPrefix$dummySha2/scripts/BUILD.bazel#L184
+      $permalinkPrefix$dummySha2/scripts/BUILD.bazel#L185
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt]: Related to an earlier comment, but I suggest having a distinct test for the sha logic vs. hiding it as part of this test.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @Sparsh1212! Took a full pass.

> @BenHenning PTAL at this issue as a POC for this script.
> Ignore the other issues on this repo as they were for debugging purposes.
> [Sparsh1212/oppia-android-clone#4](https://github.com/Sparsh1212/oppia-android-clone/issues/4)
> 
> TLDR: Actual behavior was the same as the expected behavior for all cases.
> 
> The flow of testing the script was as follows:
> 
> * At first, there were no todos in the codebase for this issue, so when I closed this, this wasn't reopened and no comment was posted.
> * Then I reopened it manually to check if the workflow passes and nothing else happens, the actual behavior was the same as expected.
> * Then I added some todos in the repo corresponding to this issue via a commit and then closed this issue. The actual result was: The issue was reopened automatically by the GitHub actions bot and a comment was also posted mentioning the TODO list.
> * I again closed the issue to check if a duplicate comment is being posted or not, and the result was that now it was gained reopened by the action bot but this time no comment was added.
> * The final step was to add more commits to the codebase which adds more todos for this issue. I again closed the issue, the result was it got reopened and a new comment was added since the todo list got changed.

I think the behavior demonstrated in your test repo looks really good. I have nothing to add that we haven't already discussed in meetings.

> @BenHenning
> I have not added the `ClosedIssueCheckTest` todo exemptions to the exemptions list yet, as the PR is in review for the first time. Will update it after the first pass. So the TODO failing CI is actually expected.

Could you resolve this now that the PR has gone through a full review round? I'd prefer to get all relevant CI checks passing vs. having to do that at the end when everything else is addressed.

File: .github/workflows/closed_issue_check.yml (hunk 1 above)
Comment (Sparsh1212) [reply .github/workflows/closed_issue_check.yml]: Done.

File: .github/workflows/closed_issue_check.yml (hunk 2 above)
Comment (Sparsh1212) [reply .github/workflows/closed_issue_check.yml]: Done.

File: .github/workflows/closed_issue_check.yml (hunk 3 above)
Comment (Sparsh1212) [reply .github/workflows/closed_issue_check.yml]: Corrected.

File: scripts/BUILD.bazel (hunk 4 above)
Comment (Sparsh1212) [reply scripts/BUILD.bazel]: Done.

File: scripts/BUILD.bazel (hunk 5 above)
Comment (Sparsh1212) [reply scripts/BUILD.bazel]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt (hunk 6 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt]: We are using a 40 char hash in the workflow, so I think it would be really better if we also use it here too.
Done based on this.

File: scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt (hunk 18 above)
Comment (Sparsh1212) [reply scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt (hunk 17 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt]: Added.

File: scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt (hunk 16 above)
Comment (Sparsh1212) [reply scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt]: Thanks.

File: scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt (hunk 15 above)
Comment (Sparsh1212) [reply scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt]: Ah! I tried and it looks like it messes up the automatic previewing.

File: scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt (hunk 14 above)
Comment (Sparsh1212) [reply scripts/src/javatests/org/oppia/android/scripts/todo/CommentCheckTest.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt (hunk 7 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt (hunk 8 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/ClosedIssueCheck.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt (hunk 9 above)
Comment (Sparsh1212) [reply scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt (hunk 10 above)
Comment (Sparsh1212) [reply scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt (hunk 11 above)
Comment (Sparsh1212) [reply scripts/src/javatests/org/oppia/android/scripts/todo/ClosedIssueCheckTest.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt (hunk 13 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt]: Done.

**Comment by Sparsh1212** [comment]
Thanks, @BenHenning 
PTAL.

File: scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt (hunk 12 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt]: Sorry I din't catch this one. Can you please explain it a bit. Currently if this fails then we are not adding a new comment in the workflow.

**Comment by Sparsh1212** [comment]
@vinitamurthi Can you please take a look?

File: .github/workflows/issue_checks.yml [hunk 19]
```diff
@@ -11,7 +11,7 @@ permissions:
 
 jobs:
   script_check:
-    name: Script Check
+    name: Closed Issue Check
```
Comment (BenHenning) [inline .github/workflows/issue_checks.yml]:
```suggestion
    name: Closed TODO Issue Check
```

(Let's try to be as specific as we can in this context).

File: scripts/src/java/org/oppia/android/scripts/todo/IssueTodosCommentCheck.kt [hunk 20]
```diff
@@ -29,6 +29,11 @@ fun main(vararg args: String) {
 
   val failureCommentFilePath = args[2]
 
+  val permaLinkPrefix = "https://github.com/oppia/oppia-android/blob/"
+
+  // Here, we are adding 40 to account for the commit Sha size
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/IssueTodosCommentCheck.kt]:
```suggestion
  // Here, we are adding 40 to account for the commit SHA-1 hash length (for context: https://en.wikipedia.org/wiki/SHA-1).
```

This helps clarify the correctness behind the magic number.

File: scripts/src/java/org/oppia/android/scripts/todo/IssueTodosResolvedCheck.kt [hunk 21]
```diff
@@ -0,0 +1,106 @@
...
+ * @param githubPermalinkUrl the GitHub url for the permalinks
+ */
+private fun generateTodoListFile(
+  repoPath: String,
+  issueTodosResolvedFailures: List<Todo>,
+  githubPermalinkUrl: String
+) {
+  val todoListFile = File(repoPath + "script_failures.txt")
+  todoListFile.appendText("The issue is reopened because of the following unresolved TODOs:")
+  todoListFile.appendText("\n")
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/IssueTodosResolvedCheck.kt]:
```suggestion
  todoListFile.appendText("The issue is reopened because of the following unresolved TODOs:\n")
```

Can be done in one line. Ditto below.

File: scripts/src/java/org/oppia/android/scripts/todo/IssueTodosResolvedCheck.kt [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+ *   bazel run //scripts:issue_todos_resolved_check -- <path_to_directory_root>
+ *   <closed_issue_number> <github_sha>
+ *
+ * Arguments:
+ * - path_to_directory_root: directory path to the root of the Oppia Android repository.
+ * - closed_issue_number: issue number of the closed issue.
+ * - github_sha: sha of the latest commit on the develop branch.
+ *
+ * Example:
+ *   bazel run //scripts:issue_todos_resolved_check -- $(pwd) 6
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/IssueTodosResolvedCheck.kt]: Nit: break at the space instead of the first digit of the hash.

File: scripts/src/javatests/org/oppia/android/scripts/todo/IssueTodosCommentCheckTest.kt [hunk 23]
```diff
@@ -0,0 +1,120 @@
...
+  private val dummySha1 = "51ab6a0341cfb86d95a387438fc993b5eb977b83"
+  private val dummySha2 = "74cd6a0341cfb86d95a387438fc993b5eb977b83"
+
+  @Rule
+  @JvmField
+  var tempFolder = TemporaryFolder()
+
+  /**
+   * This is the case when no comment is present on the issue. In this case the Github API results
+   * in an empty comment body.
```
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/todo/IssueTodosCommentCheckTest.kt]:
```suggestion
   * This is the case when no comment is present on the issue. In this case the GitHub API produces
   * an empty comment body.
```

File: scripts/src/java/org/oppia/android/scripts/todo/IssueTodosCommentCheck.kt [hunk 24]
```diff
@@ -3,11 +3,11 @@ package org.oppia.android.scripts.todo
 import java.io.File
 
 /**
- * Script for ensuring that the [ClosedIssueCheck] failure comment is not the same as the latest
- * comment of the closed issue.
+ * Script for ensuring that the [IssueTodosResolvedCheck] failure comment is not the same as the
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/IssueTodosCommentCheck.kt]: The new name actually seems backwards to me. What do you think about 'TodoIssueResolvedCheck' instead? Ditto for the other one & all corresponding doc comments and BUILD targets.

File: scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt (hunk 12 above)
Comment (BenHenning) [reply scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt]: As discussed over chat, what I mean is that it seems like checks should pass if their expected state is achieved (in this case, that the comment matches & no action needs to take place). Failures should occur when the non-default state is encountered (i.e. that a new comment is needed). Reversing this pass/fail in the script would be more predictable behavior, I think.

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @Sparsh1212. Generally LGTM, just have a few comments.

Also, please update the title to include the base PR that this is blocked on. I can't approve this until that PR is merged. Please assign back once the comments below are addressed & the base PR is merged.

File: .github/workflows/issue_checks.yml (hunk 19 above)
Comment (Sparsh1212) [reply .github/workflows/issue_checks.yml]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/IssueTodosCommentCheck.kt (hunk 20 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/IssueTodosCommentCheck.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/IssueTodosResolvedCheck.kt (hunk 21 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/IssueTodosResolvedCheck.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/IssueTodosResolvedCheck.kt (hunk 22 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/IssueTodosResolvedCheck.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/todo/IssueTodosCommentCheckTest.kt (hunk 23 above)
Comment (Sparsh1212) [reply scripts/src/javatests/org/oppia/android/scripts/todo/IssueTodosCommentCheckTest.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/IssueTodosCommentCheck.kt (hunk 24 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/IssueTodosCommentCheck.kt]: Sounds good. Done.

File: scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt (hunk 12 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/CommentCheck.kt]: Done.

File: scripts/src/java/org/oppia/android/scripts/todo/TodoIssueResolvedCheck.kt [hunk 25]
```diff
@@ -0,0 +1,105 @@
...
+ *   <closed_issue_number> <github_sha>
+ *
+ * Arguments:
+ * - path_to_directory_root: directory path to the root of the Oppia Android repository.
+ * - closed_issue_number: issue number of the closed issue.
+ * - github_sha: sha of the latest commit on the develop branch.
+ *
+ * Example:
+ *   bazel run //scripts:todo_issue_resolved_check -- $(pwd) 6
+ *   77ff8361b4bde52f695ceb91aa1aab36932a94fe
```
Comment (BenHenning) [inline scripts/src/java/org/oppia/android/scripts/todo/TodoIssueResolvedCheck.kt]:
```suggestion
 *   bazel run //scripts:todo_issue_resolved_check -- $(pwd)
 *   677ff8361b4bde52f695ceb91aa1aab36932a94fe
```

(prev comment wasn't resolved).

**Review by BenHenning — COMMENTED** [review]
Just one comment, otherwise LGTM @Sparsh1212.

File: scripts/src/java/org/oppia/android/scripts/todo/TodoIssueResolvedCheck.kt (hunk 25 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/TodoIssueResolvedCheck.kt]: Ah! 6 is actuay the example issue number.

**Comment by Sparsh1212** [comment]
> Just one comment, otherwise LGTM @Sparsh1212.

@BenHenning  Followed up. PTAL.

File: scripts/src/java/org/oppia/android/scripts/todo/TodoIssueResolvedCheck.kt (hunk 25 above)
Comment (Sparsh1212) [reply scripts/src/java/org/oppia/android/scripts/todo/TodoIssueResolvedCheck.kt]: Now, I have broken this down before 6 so as to avoid confusion. PTAL.

**Review by BenHenning — COMMENTED** [review]
Thanks @Sparsh1212. This LGTM, please assign back once #3508 is merged & this PR is updated.

**Review by BenHenning — APPROVED** [review]
Thanks @Sparsh1212. LGTM!
